src/instructlab/model/train.py

```diff
@@ -520,20 +580,340 @@ def get_files(directory: str, pattern: str) -> list[str]:
+
+    from instructlab.model.evaluate import launch_server
+    import torch
+
+    ckpt_path = str(ckpt_path)
+    eval_cache_path = str(eval_cache_path)
+
+    # hard-override for local insecure vLLM serving
+    ctx.params["tls_client_cert"] = None
+    ctx.params["tls_client_key"] = None
+    ctx.params["tls_client_passwd"] = None
+    ctx.params["tls_insecure"] = True
+
+    evaluator = MTBenchEvaluator(
+        model_name=ckpt_path,
+        judge_model_name=judge_name,
+        output_dir=eval_cache_path,
+        merge_system_user_message=True,
+    )
+
+    server = None
+    model_serve_url = None
+    try:
+        logger.debug("Starting model server for mt-bench answer generation")
+        server, model_serve_url = launch_server(
+            ctx=ctx,
+            model=ckpt_path,
+            model_name=ckpt_path,
+            gpus=torch.cuda.device_count(),
+            max_workers=None,  # not needed for vLLM
+            enable_serving_output=False,
+            backend="vllm",
+        )
+        logger.debug("Generating mt-bench answers")
+        evaluator.gen_answers(model_serve_url)
+    finally:
+        if server is not None:
+            server.shutdown()
+
+    ckpt_score: typing.Optional[int] = None
+    try:
+        logger.debug("Starting model server for mt-bench answer judgement")
+        # TODO: remove hard-code model path
+        server, model_serve_url = launch_server(
+            ctx=ctx,
+            model=judge_name,
+            model_name=judge_name,
+            gpus=torch.cuda.device_count(),
+            max_workers=None,  # not needed for vLLM
+            backend="vllm",
+            enable_serving_output=False,
+        )
+        logger.debug("Judging mt-bench answers")
+        ckpt_score, _, _, _ = evaluator.judge_answers(model_serve_url)
+    finally:
+        if server is not None:
+            server.shutdown()
+
+    return ckpt_score
+
+
+def _evaluate_dir_of_checkpoints(
+    checkpoints_dir_path: str, eval_func: typing.Callable[..., float]
+) -> str:
+    """Run eval_func on all model checkpoints in a directory, returning the best performing."""
+
+    # TODO: parallelize MMLU over available GPUs
+
+    results: list[typing.Tuple[float, str]] = []
+    directory = Path(checkpoints_dir_path)
+    for ckpt_path in directory.iterdir():
+        if ckpt_path.is_dir():
+            logger.debug(ckpt_path)
+            ckpt_score = eval_func(ckpt_path=ckpt_path)
             click.secho(
-                f"Error while executing training library {exc}",
+                f"CHECKPOINT EVALUATION: {ckpt_path} SCORED {ckpt_score}",
                 fg="red",
+                bg="cyan",
             )
-            raise click.exceptions.Exit(1)
+            results.append((ckpt_score, ckpt_path))
+
+    # shape of: [tuple(score, path)]
+    ckpt_performance = sorted(results, reverse=True)
+
+    # (score, path)
+    return ckpt_performance[0]
```

Comment:
Should we verify that this list always has a length of at least 1? I'm worried about the case when the checkpoints directory is empty and we fail to obtain any evaluation scores.

Reply:
ack